### PR TITLE
Implements init_dataframe as multiple codegen functions

### PR DESCRIPTION
**Pull request by kozlov-alexey**
Motivation: init_dataframe was implemented via Numba intrinsic taking *args,
which seems to generate redundant extractvalue/insertvalue LLVM
instructions, producing quadratic IR when number of DF columns grows and affecting
total compilation time of function that create large DFs. This PR
replaces singe init_dataframe with multiple functions basing on number of columns
in a DF which are generated at compile time, thus avoiding use of *args.

n_columns |   | 8 | 16 | 32 | 64 | 128 | 256 | 512
-- | -- | -- | -- | -- | -- | -- | -- | --
LLVM IR size, Mb | on master | 0.287622 | 0.55394 | 1.262865 | 3.383549 | 10.44003 | 35.79943 | 131.384
LLVM IR size, Mb | With PR #936 | 0.143275 | 0.209119 | 0.341938 | 0.608992 | 1.143528 | 2.220672 | 4.406426
ratio without/with |   | 2.007482 | 2.648924 | 3.693257 | 5.555986 | 9.12967 | 16.12099 | 29.81645
compilation time, s | on master | 0.521313 | 0.366884 | 0.67621 | 1.39326 | 4.603106 | 17.54948 | 126.7943
compilation time, s | With PR #936 | 0.683099 | 0.413965 | 0.450348 | 0.715598 | 1.454044 | 3.210638 | 6.943996
ratio without/with |   | 0.763159 | 0.886268 | 1.501529 | 1.946987 | 3.165726 | 5.466041 | 18.25956

### CODE REVIEW

**Comment by kozlov-alexey** [comment]
Test failures of read_csv tests with:

> Failed in nopython mode pipeline (step: nopython rewrites)
>            module 'sdc.hiframes.pd_dataframe_ext' has no attribute 'init_dataframe'
> 
are expected because this PR requires changes from #918 which was rolled-back recently. So this will be blocked until #918 is returned.

File: sdc/rewrites/dataframe_constructor.py [hunk 1]
```diff
@@ -79,18 +87,45 @@ def match(self, func_ir, block, typemap, calltypes):
...
             args = get_call_parameters(call=stmt.value, arg_names=self._df_arg_list)
-
             old_data = args['data']
-
             args['data'], args['columns'] = self._extract_dict_args(args, self._func_ir)
 
+            args_len = len(args['data'])
+            func_name = f'init_dataframe_{args_len}'
+
+            injected_module = modules[pd_dataframe_ext_module.__name__]
```
Comment (AlexanderKalistratov) [inline sdc/rewrites/dataframe_constructor.py:98]: Why do you need this? Isn't 
```python
pd_dataframe_ext_module == modules[pd_dataframe_ext_module.__name__]
```
?

File: sdc/rewrites/dataframe_constructor.py [hunk 2]
```diff
@@ -130,42 +165,118 @@ def _replace_call(stmt, new_call, args, block, func_ir):
...
+    params = ', '.join(args_col_data + ['index'] + args_col_names)
+    suffix = ('' if n_cols == 0 else ', ')
+
+    func_text = dedent(
+        f'''
+    @intrinsic
+    def {func_name}(typingctx, {params}):
+        """Create a DataFrame with provided data, index and columns values.
+        Used as a single constructor for DataFrame and assigning its data, so that
+        optimization passes can look for init_dataframe() to see if underlying
```
Comment (AlexanderKalistratov) [inline sdc/rewrites/dataframe_constructor.py:192]: I think this comment is outdated and could confuse

**Comment by AlexanderKalistratov** [comment]
@kozlov-alexey @xaleryb win 3.6 build fails with svml error again:
```
test_series_apply_np (sdc.tests.test_series.TestSeries) ... LLVM ERROR: Symbol not found: __svml_log4_ha
```

File: sdc/rewrites/dataframe_constructor.py (hunk 1 above)
Comment (kozlov-alexey) [reply sdc/rewrites/dataframe_constructor.py]: Right, thanks.

File: sdc/rewrites/dataframe_constructor.py (hunk 2 above)
Comment (kozlov-alexey) [reply sdc/rewrites/dataframe_constructor.py]: Yeap, updated the docstring.

**Comment by kozlov-alexey** [comment]
> @kozlov-alexey @xaleryb win 3.6 build fails with svml error again:
> 
> ```
> test_series_apply_np (sdc.tests.test_series.TestSeries) ... LLVM ERROR: Symbol not found: __svml_log4_ha
> ```

I think something's wrong with the packages being used (see mkl and many others are installed from public channels, but not built). Can this be a reason?

**Review by AlexanderKalistratov — APPROVED** [review]

**Pull request merged**